--!strict

local native = require("@self/native")
local raylib = require("@self/native/raylib")

export type Vector2 = raylib.Vector2
export type Vector3 = raylib.Vector3
export type Vector4 = raylib.Vector4
export type Quaternion = raylib.Quaternion
export type Matrix = raylib.Matrix
export type Color = raylib.Color
export type Rectangle = raylib.Rectangle
export type Image = raylib.Image
export type Texture = raylib.Texture
export type Texture2D = raylib.Texture2D
export type TextureCubemap = raylib.TextureCubemap
export type RenderTexture = raylib.RenderTexture
export type RenderTexture2D = raylib.RenderTexture2D
export type NPatchInfo = raylib.NPatchInfo
export type GlyphInfo = raylib.GlyphInfo
export type Font = raylib.Font
export type Camera3D = raylib.Camera3D
export type Camera2D = raylib.Camera2D
export type Mesh = raylib.Mesh
export type Shader = raylib.Shader
export type MaterialMap = raylib.MaterialMap
export type Material = raylib.Material
export type Transform = raylib.Transform
export type BoneInfo = raylib.BoneInfo
export type Model = raylib.Model
export type ModelAnimation = raylib.ModelAnimation
export type Ray = raylib.Ray
export type RayCollision = raylib.RayCollision
export type BoundingBox = raylib.BoundingBox
export type Wave = raylib.Wave
export type rAudioBuffer = raylib.rAudioBuffer
export type rAudioProcessor = raylib.rAudioProcessor
export type AudioStream = raylib.AudioStream
export type Sound = raylib.Sound
export type Music = raylib.Music
export type VrDeviceInfo = raylib.VrDeviceInfo
export type VrStereoConfig = raylib.VrStereoConfig
export type FilePathList = raylib.FilePathList
export type AutomationEvent = raylib.AutomationEvent
export type AutomationEventList = raylib.AutomationEventList

local lib = native.lib
local const = native.const

return {
	AttachAudioMixedProcessor = lib.AttachAudioMixedProcessor,
	AttachAudioStreamProcessor = lib.AttachAudioStreamProcessor,
	BeginBlendMode = lib.BeginBlendMode,
	BeginDrawing = lib.BeginDrawing,
	BeginMode2D = lib.BeginMode2D,
	BeginMode3D = lib.BeginMode3D,
	BeginScissorMode = lib.BeginScissorMode,
	BeginShaderMode = lib.BeginShaderMode,
	BeginTextureMode = lib.BeginTextureMode,
	BeginVrStereoMode = lib.BeginVrStereoMode,
	BuildPoseFromParentJoints = lib.BuildPoseFromParentJoints,
	CameraMoveForward = lib.CameraMoveForward,
	CameraMoveRight = lib.CameraMoveRight,
	CameraMoveToTarget = lib.CameraMoveToTarget,
	CameraMoveUp = lib.CameraMoveUp,
	CameraPitch = lib.CameraPitch,
	CameraRoll = lib.CameraRoll,
	CameraYaw = lib.CameraYaw,
	ChangeDirectory = lib.ChangeDirectory,
	CharCallback = lib.CharCallback,
	CheckCollisionBoxSphere = lib.CheckCollisionBoxSphere,
	CheckCollisionBoxes = lib.CheckCollisionBoxes,
	CheckCollisionCircleLine = lib.CheckCollisionCircleLine,
	CheckCollisionCircleRec = lib.CheckCollisionCircleRec,
	CheckCollisionCircles = lib.CheckCollisionCircles,
	CheckCollisionLines = lib.CheckCollisionLines,
	CheckCollisionPointCircle = lib.CheckCollisionPointCircle,
	CheckCollisionPointLine = lib.CheckCollisionPointLine,
	CheckCollisionPointPoly = lib.CheckCollisionPointPoly,
	CheckCollisionPointRec = lib.CheckCollisionPointRec,
	CheckCollisionPointTriangle = lib.CheckCollisionPointTriangle,
	CheckCollisionRecs = lib.CheckCollisionRecs,
	CheckCollisionSpheres = lib.CheckCollisionSpheres,
	Clamp = lib.Clamp,
	ClearBackground = lib.ClearBackground,
	ClearWindowState = lib.ClearWindowState,
	CloseAudioDevice = lib.CloseAudioDevice,
	ClosePlatform = lib.ClosePlatform,
	CloseWindow = lib.CloseWindow,
	CodepointToUTF8 = lib.CodepointToUTF8,
	ColorAlpha = lib.ColorAlpha,
	ColorAlphaBlend = lib.ColorAlphaBlend,
	ColorBrightness = lib.ColorBrightness,
	ColorContrast = lib.ColorContrast,
	ColorFromHSV = lib.ColorFromHSV,
	ColorFromNormalized = lib.ColorFromNormalized,
	ColorIsEqual = lib.ColorIsEqual,
	ColorLerp = lib.ColorLerp,
	ColorNormalize = lib.ColorNormalize,
	ColorTint = lib.ColorTint,
	ColorToHSV = lib.ColorToHSV,
	ColorToInt = lib.ColorToInt,
	CompressData = lib.CompressData,
	ComputeCRC32 = lib.ComputeCRC32,
	ComputeMD5 = lib.ComputeMD5,
	ComputeSHA1 = lib.ComputeSHA1,
	CursorEnterCallback = lib.CursorEnterCallback,
	DecodeDataBase64 = lib.DecodeDataBase64,
	DecompressData = lib.DecompressData,
	DetachAudioMixedProcessor = lib.DetachAudioMixedProcessor,
	DetachAudioStreamProcessor = lib.DetachAudioStreamProcessor,
	DirectoryExists = lib.DirectoryExists,
	DisableCursor = lib.DisableCursor,
	DisableEventWaiting = lib.DisableEventWaiting,
	DrawBillboard = lib.DrawBillboard,
	DrawBillboardPro = lib.DrawBillboardPro,
	DrawBillboardRec = lib.DrawBillboardRec,
	DrawBoundingBox = lib.DrawBoundingBox,
	DrawCapsule = lib.DrawCapsule,
	DrawCapsuleWires = lib.DrawCapsuleWires,
	DrawCircle = lib.DrawCircle,
	DrawCircle3D = lib.DrawCircle3D,
	DrawCircleGradient = lib.DrawCircleGradient,
	DrawCircleLines = lib.DrawCircleLines,
	DrawCircleLinesV = lib.DrawCircleLinesV,
	DrawCircleSector = lib.DrawCircleSector,
	DrawCircleSectorLines = lib.DrawCircleSectorLines,
	DrawCircleV = lib.DrawCircleV,
	DrawCube = lib.DrawCube,
	DrawCubeV = lib.DrawCubeV,
	DrawCubeWires = lib.DrawCubeWires,
	DrawCubeWiresV = lib.DrawCubeWiresV,
	DrawCylinder = lib.DrawCylinder,
	DrawCylinderEx = lib.DrawCylinderEx,
	DrawCylinderWires = lib.DrawCylinderWires,
	DrawCylinderWiresEx = lib.DrawCylinderWiresEx,
	DrawEllipse = lib.DrawEllipse,
	DrawEllipseLines = lib.DrawEllipseLines,
	DrawFPS = lib.DrawFPS,
	DrawGrid = lib.DrawGrid,
	DrawLine = lib.DrawLine,
	DrawLine3D = lib.DrawLine3D,
	DrawLineBezier = lib.DrawLineBezier,
	DrawLineEx = lib.DrawLineEx,
	DrawLineStrip = lib.DrawLineStrip,
	DrawLineV = lib.DrawLineV,
	DrawMesh = lib.DrawMesh,
	DrawMeshInstanced = lib.DrawMeshInstanced,
	DrawModel = lib.DrawModel,
	DrawModelEx = lib.DrawModelEx,
	DrawModelPoints = lib.DrawModelPoints,
	DrawModelPointsEx = lib.DrawModelPointsEx,
	DrawModelWires = lib.DrawModelWires,
	DrawModelWiresEx = lib.DrawModelWiresEx,
	DrawPixel = lib.DrawPixel,
	DrawPixelV = lib.DrawPixelV,
	DrawPlane = lib.DrawPlane,
	DrawPoint3D = lib.DrawPoint3D,
	DrawPoly = lib.DrawPoly,
	DrawPolyLines = lib.DrawPolyLines,
	DrawPolyLinesEx = lib.DrawPolyLinesEx,
	DrawRay = lib.DrawRay,
	DrawRectangle = lib.DrawRectangle,
	DrawRectangleGradientEx = lib.DrawRectangleGradientEx,
	DrawRectangleGradientH = lib.DrawRectangleGradientH,
	DrawRectangleGradientV = lib.DrawRectangleGradientV,
	DrawRectangleLines = lib.DrawRectangleLines,
	DrawRectangleLinesEx = lib.DrawRectangleLinesEx,
	DrawRectanglePro = lib.DrawRectanglePro,
	DrawRectangleRec = lib.DrawRectangleRec,
	DrawRectangleRounded = lib.DrawRectangleRounded,
	DrawRectangleRoundedLines = lib.DrawRectangleRoundedLines,
	DrawRectangleRoundedLinesEx = lib.DrawRectangleRoundedLinesEx,
	DrawRectangleV = lib.DrawRectangleV,
	DrawRing = lib.DrawRing,
	DrawRingLines = lib.DrawRingLines,
	DrawSphere = lib.DrawSphere,
	DrawSphereEx = lib.DrawSphereEx,
	DrawSphereWires = lib.DrawSphereWires,
	DrawSplineBasis = lib.DrawSplineBasis,
	DrawSplineBezierCubic = lib.DrawSplineBezierCubic,
	DrawSplineBezierQuadratic = lib.DrawSplineBezierQuadratic,
	DrawSplineCatmullRom = lib.DrawSplineCatmullRom,
	DrawSplineLinear = lib.DrawSplineLinear,
	DrawSplineSegmentBasis = lib.DrawSplineSegmentBasis,
	DrawSplineSegmentBezierCubic = lib.DrawSplineSegmentBezierCubic,
	DrawSplineSegmentBezierQuadratic = lib.DrawSplineSegmentBezierQuadratic,
	DrawSplineSegmentCatmullRom = lib.DrawSplineSegmentCatmullRom,
	DrawSplineSegmentLinear = lib.DrawSplineSegmentLinear,
	DrawText = lib.DrawText,
	DrawTextCodepoint = lib.DrawTextCodepoint,
	DrawTextCodepoints = lib.DrawTextCodepoints,
	DrawTextEx = lib.DrawTextEx,
	DrawTextPro = lib.DrawTextPro,
	DrawTexture = lib.DrawTexture,
	DrawTextureEx = lib.DrawTextureEx,
	DrawTextureNPatch = lib.DrawTextureNPatch,
	DrawTexturePro = lib.DrawTexturePro,
	DrawTextureRec = lib.DrawTextureRec,
	DrawTextureV = lib.DrawTextureV,
	DrawTriangle = lib.DrawTriangle,
	DrawTriangle3D = lib.DrawTriangle3D,
	DrawTriangleFan = lib.DrawTriangleFan,
	DrawTriangleLines = lib.DrawTriangleLines,
	DrawTriangleStrip = lib.DrawTriangleStrip,
	DrawTriangleStrip3D = lib.DrawTriangleStrip3D,
	EnableCursor = lib.EnableCursor,
	EnableEventWaiting = lib.EnableEventWaiting,
	EncodeDataBase64 = lib.EncodeDataBase64,
	EndBlendMode = lib.EndBlendMode,
	EndDrawing = lib.EndDrawing,
	EndMode2D = lib.EndMode2D,
	EndMode3D = lib.EndMode3D,
	EndScissorMode = lib.EndScissorMode,
	EndShaderMode = lib.EndShaderMode,
	EndTextureMode = lib.EndTextureMode,
	EndVrStereoMode = lib.EndVrStereoMode,
	ErrorCallback = lib.ErrorCallback,
	ExportAutomationEventList = lib.ExportAutomationEventList,
	ExportDataAsCode = lib.ExportDataAsCode,
	ExportFontAsCode = lib.ExportFontAsCode,
	ExportImage = lib.ExportImage,
	ExportImageAsCode = lib.ExportImageAsCode,
	ExportImageToMemory = lib.ExportImageToMemory,
	ExportMesh = lib.ExportMesh,
	ExportMeshAsCode = lib.ExportMeshAsCode,
	ExportWave = lib.ExportWave,
	ExportWaveAsCode = lib.ExportWaveAsCode,
	Fade = lib.Fade,
	FileExists = lib.FileExists,
	FloatEquals = lib.FloatEquals,
	FloatToHalf = lib.FloatToHalf,
	GenImageCellular = lib.GenImageCellular,
	GenImageChecked = lib.GenImageChecked,
	GenImageColor = lib.GenImageColor,
	GenImageFontAtlas = lib.GenImageFontAtlas,
	GenImageGradientLinear = lib.GenImageGradientLinear,
	GenImageGradientRadial = lib.GenImageGradientRadial,
	GenImageGradientSquare = lib.GenImageGradientSquare,
	GenImagePerlinNoise = lib.GenImagePerlinNoise,
	GenImageText = lib.GenImageText,
	GenImageWhiteNoise = lib.GenImageWhiteNoise,
	GenMeshCone = lib.GenMeshCone,
	GenMeshCube = lib.GenMeshCube,
	GenMeshCubicmap = lib.GenMeshCubicmap,
	GenMeshCylinder = lib.GenMeshCylinder,
	GenMeshHeightmap = lib.GenMeshHeightmap,
	GenMeshHemiSphere = lib.GenMeshHemiSphere,
	GenMeshKnot = lib.GenMeshKnot,
	GenMeshPlane = lib.GenMeshPlane,
	GenMeshPoly = lib.GenMeshPoly,
	GenMeshSphere = lib.GenMeshSphere,
	GenMeshTangents = lib.GenMeshTangents,
	GenMeshTorus = lib.GenMeshTorus,
	GenTextureMipmaps = lib.GenTextureMipmaps,
	GetApplicationDirectory = lib.GetApplicationDirectory,
	GetCameraForward = lib.GetCameraForward,
	GetCameraMatrix = lib.GetCameraMatrix,
	GetCameraMatrix2D = lib.GetCameraMatrix2D,
	GetCameraProjectionMatrix = lib.GetCameraProjectionMatrix,
	GetCameraRight = lib.GetCameraRight,
	GetCameraUp = lib.GetCameraUp,
	GetCameraViewMatrix = lib.GetCameraViewMatrix,
	GetCharPressed = lib.GetCharPressed,
	GetClipboardImage = lib.GetClipboardImage,
	GetClipboardText = lib.GetClipboardText,
	GetCodepoint = lib.GetCodepoint,
	GetCodepointCount = lib.GetCodepointCount,
	GetCodepointNext = lib.GetCodepointNext,
	GetCodepointPrevious = lib.GetCodepointPrevious,
	GetCollisionRec = lib.GetCollisionRec,
	GetColor = lib.GetColor,
	GetCurrentMonitor = lib.GetCurrentMonitor,
	GetDirectoryPath = lib.GetDirectoryPath,
	GetFPS = lib.GetFPS,
	GetFileExtension = lib.GetFileExtension,
	GetFileLength = lib.GetFileLength,
	GetFileModTime = lib.GetFileModTime,
	GetFileName = lib.GetFileName,
	GetFileNameWithoutExt = lib.GetFileNameWithoutExt,
	GetFontDefault = lib.GetFontDefault,
	GetFrameTime = lib.GetFrameTime,
	GetGamepadAxisCount = lib.GetGamepadAxisCount,
	GetGamepadAxisMovement = lib.GetGamepadAxisMovement,
	GetGamepadButtonPressed = lib.GetGamepadButtonPressed,
	GetGamepadName = lib.GetGamepadName,
	GetGestureDetected = lib.GetGestureDetected,
	GetGestureDragAngle = lib.GetGestureDragAngle,
	GetGestureDragVector = lib.GetGestureDragVector,
	GetGestureHoldDuration = lib.GetGestureHoldDuration,
	GetGesturePinchAngle = lib.GetGesturePinchAngle,
	GetGesturePinchVector = lib.GetGesturePinchVector,
	GetGlyphAtlasRec = lib.GetGlyphAtlasRec,
	GetGlyphIndex = lib.GetGlyphIndex,
	GetGlyphInfo = lib.GetGlyphInfo,
	GetImageAlphaBorder = lib.GetImageAlphaBorder,
	GetImageColor = lib.GetImageColor,
	GetKeyName = lib.GetKeyName,
	GetKeyPressed = lib.GetKeyPressed,
	GetLine = lib.GetLine,
	GetMasterVolume = lib.GetMasterVolume,
	GetMeshBoundingBox = lib.GetMeshBoundingBox,
	GetModelBoundingBox = lib.GetModelBoundingBox,
	rlSetLineWidth = lib.rlSetLineWidth,
	GetMonitorCount = lib.GetMonitorCount,
	GetMonitorHeight = lib.GetMonitorHeight,
	GetMonitorName = lib.GetMonitorName,
	GetMonitorPhysicalHeight = lib.GetMonitorPhysicalHeight,
	GetMonitorPhysicalWidth = lib.GetMonitorPhysicalWidth,
	GetMonitorPosition = lib.GetMonitorPosition,
	GetMonitorRefreshRate = lib.GetMonitorRefreshRate,
	GetMonitorWidth = lib.GetMonitorWidth,
	GetMouseDelta = lib.GetMouseDelta,
	GetMousePosition = lib.GetMousePosition,
	GetMouseWheelMove = lib.GetMouseWheelMove,
	GetMouseWheelMoveV = lib.GetMouseWheelMoveV,
	GetMouseX = lib.GetMouseX,
	GetMouseY = lib.GetMouseY,
	GetMusicTimeLength = lib.GetMusicTimeLength,
	GetMusicTimePlayed = lib.GetMusicTimePlayed,
	GetPixelColor = lib.GetPixelColor,
	GetPixelDataSize = lib.GetPixelDataSize,
	GetPoseAtTimeGLTF = lib.GetPoseAtTimeGLTF,
	GetPrevDirectoryPath = lib.GetPrevDirectoryPath,
	GetRandomValue = lib.GetRandomValue,
	GetRayCollisionBox = lib.GetRayCollisionBox,
	GetRayCollisionMesh = lib.GetRayCollisionMesh,
	GetRayCollisionQuad = lib.GetRayCollisionQuad,
	GetRayCollisionSphere = lib.GetRayCollisionSphere,
	GetRayCollisionTriangle = lib.GetRayCollisionTriangle,
	GetRenderHeight = lib.GetRenderHeight,
	GetRenderWidth = lib.GetRenderWidth,
	GetScreenHeight = lib.GetScreenHeight,
	GetScreenToWorld2D = lib.GetScreenToWorld2D,
	GetScreenToWorldRay = lib.GetScreenToWorldRay,
	GetScreenToWorldRayEx = lib.GetScreenToWorldRayEx,
	GetScreenWidth = lib.GetScreenWidth,
	GetShaderLocation = lib.GetShaderLocation,
	GetShaderLocationAttrib = lib.GetShaderLocationAttrib,
	GetShapesTexture = lib.GetShapesTexture,
	GetShapesTextureRectangle = lib.GetShapesTextureRectangle,
	GetSplinePointBasis = lib.GetSplinePointBasis,
	GetSplinePointBezierCubic = lib.GetSplinePointBezierCubic,
	GetSplinePointBezierQuad = lib.GetSplinePointBezierQuad,
	GetSplinePointCatmullRom = lib.GetSplinePointCatmullRom,
	GetSplinePointLinear = lib.GetSplinePointLinear,
	GetTime = lib.GetTime,
	GetTouchPointCount = lib.GetTouchPointCount,
	GetTouchPointId = lib.GetTouchPointId,
	GetTouchPosition = lib.GetTouchPosition,
	GetTouchX = lib.GetTouchX,
	GetTouchY = lib.GetTouchY,
	GetWindowHandle = lib.GetWindowHandle,
	GetWindowPosition = lib.GetWindowPosition,
	GetWindowScaleDPI = lib.GetWindowScaleDPI,
	GetWorkingDirectory = lib.GetWorkingDirectory,
	GetWorldToScreen = lib.GetWorldToScreen,
	GetWorldToScreen2D = lib.GetWorldToScreen2D,
	GetWorldToScreenEx = lib.GetWorldToScreenEx,
	HalfToFloat = lib.HalfToFloat,
	HideCursor = lib.HideCursor,
	ImageAlphaClear = lib.ImageAlphaClear,
	ImageAlphaCrop = lib.ImageAlphaCrop,
	ImageAlphaMask = lib.ImageAlphaMask,
	ImageAlphaPremultiply = lib.ImageAlphaPremultiply,
	ImageBlurGaussian = lib.ImageBlurGaussian,
	ImageClearBackground = lib.ImageClearBackground,
	ImageColorBrightness = lib.ImageColorBrightness,
	ImageColorContrast = lib.ImageColorContrast,
	ImageColorGrayscale = lib.ImageColorGrayscale,
	ImageColorInvert = lib.ImageColorInvert,
	ImageColorReplace = lib.ImageColorReplace,
	ImageColorTint = lib.ImageColorTint,
	ImageCopy = lib.ImageCopy,
	ImageCrop = lib.ImageCrop,
	ImageDither = lib.ImageDither,
	ImageDraw = lib.ImageDraw,
	ImageDrawCircle = lib.ImageDrawCircle,
	ImageDrawCircleLines = lib.ImageDrawCircleLines,
	ImageDrawCircleLinesV = lib.ImageDrawCircleLinesV,
	ImageDrawCircleV = lib.ImageDrawCircleV,
	ImageDrawLine = lib.ImageDrawLine,
	ImageDrawLineEx = lib.ImageDrawLineEx,
	ImageDrawLineV = lib.ImageDrawLineV,
	ImageDrawPixel = lib.ImageDrawPixel,
	ImageDrawPixelV = lib.ImageDrawPixelV,
	ImageDrawRectangle = lib.ImageDrawRectangle,
	ImageDrawRectangleLines = lib.ImageDrawRectangleLines,
	ImageDrawRectangleRec = lib.ImageDrawRectangleRec,
	ImageDrawRectangleV = lib.ImageDrawRectangleV,
	ImageDrawText = lib.ImageDrawText,
	ImageDrawTextEx = lib.ImageDrawTextEx,
	ImageDrawTriangle = lib.ImageDrawTriangle,
	ImageDrawTriangleEx = lib.ImageDrawTriangleEx,
	ImageDrawTriangleFan = lib.ImageDrawTriangleFan,
	ImageDrawTriangleLines = lib.ImageDrawTriangleLines,
	ImageDrawTriangleStrip = lib.ImageDrawTriangleStrip,
	ImageFlipHorizontal = lib.ImageFlipHorizontal,
	ImageFlipVertical = lib.ImageFlipVertical,
	ImageFormat = lib.ImageFormat,
	ImageFromChannel = lib.ImageFromChannel,
	ImageFromImage = lib.ImageFromImage,
	ImageKernelConvolution = lib.ImageKernelConvolution,
	ImageMipmaps = lib.ImageMipmaps,
	ImageResize = lib.ImageResize,
	ImageResizeCanvas = lib.ImageResizeCanvas,
	ImageResizeNN = lib.ImageResizeNN,
	ImageRotate = lib.ImageRotate,
	ImageRotateCCW = lib.ImageRotateCCW,
	ImageRotateCW = lib.ImageRotateCW,
	ImageText = lib.ImageText,
	ImageTextEx = lib.ImageTextEx,
	ImageToPOT = lib.ImageToPOT,
	InitAudioDevice = lib.InitAudioDevice,
	InitPlatform = lib.InitPlatform,
	InitWindow = lib.InitWindow,
	IsAudioBufferPlaying = lib.IsAudioBufferPlaying,
	IsAudioDeviceReady = lib.IsAudioDeviceReady,
	IsAudioStreamPlaying = lib.IsAudioStreamPlaying,
	IsAudioStreamProcessed = lib.IsAudioStreamProcessed,
	IsAudioStreamValid = lib.IsAudioStreamValid,
	IsCursorHidden = lib.IsCursorHidden,
	IsCursorOnScreen = lib.IsCursorOnScreen,
	IsFileDropped = lib.IsFileDropped,
	IsFileExtension = lib.IsFileExtension,
	IsFileNameValid = lib.IsFileNameValid,
	IsFontValid = lib.IsFontValid,
	IsGamepadAvailable = lib.IsGamepadAvailable,
	IsGamepadButtonDown = lib.IsGamepadButtonDown,
	IsGamepadButtonPressed = lib.IsGamepadButtonPressed,
	IsGamepadButtonReleased = lib.IsGamepadButtonReleased,
	IsGamepadButtonUp = lib.IsGamepadButtonUp,
	IsGestureDetected = lib.IsGestureDetected,
	IsImageValid = lib.IsImageValid,
	IsKeyDown = lib.IsKeyDown,
	IsKeyPressed = lib.IsKeyPressed,
	IsKeyPressedRepeat = lib.IsKeyPressedRepeat,
	IsKeyReleased = lib.IsKeyReleased,
	IsKeyUp = lib.IsKeyUp,
	IsMaterialValid = lib.IsMaterialValid,
	IsModelAnimationValid = lib.IsModelAnimationValid,
	IsModelValid = lib.IsModelValid,
	IsMouseButtonDown = lib.IsMouseButtonDown,
	IsMouseButtonPressed = lib.IsMouseButtonPressed,
	IsMouseButtonReleased = lib.IsMouseButtonReleased,
	IsMouseButtonUp = lib.IsMouseButtonUp,
	IsMusicStreamPlaying = lib.IsMusicStreamPlaying,
	IsMusicValid = lib.IsMusicValid,
	IsPathFile = lib.IsPathFile,
	IsRenderTextureValid = lib.IsRenderTextureValid,
	IsShaderValid = lib.IsShaderValid,
	IsSoundPlaying = lib.IsSoundPlaying,
	IsSoundValid = lib.IsSoundValid,
	IsTextureValid = lib.IsTextureValid,
	IsWaveValid = lib.IsWaveValid,
	IsWindowFocused = lib.IsWindowFocused,
	IsWindowFullscreen = lib.IsWindowFullscreen,
	IsWindowHidden = lib.IsWindowHidden,
	IsWindowMaximized = lib.IsWindowMaximized,
	IsWindowMinimized = lib.IsWindowMinimized,
	IsWindowReady = lib.IsWindowReady,
	IsWindowResized = lib.IsWindowResized,
	IsWindowState = lib.IsWindowState,
	JoystickCallback = lib.JoystickCallback,
	KeyCallback = lib.KeyCallback,
	Lerp = lib.Lerp,
	LoadAudioBuffer = lib.LoadAudioBuffer,
	LoadAudioStream = lib.LoadAudioStream,
	LoadAutomationEventList = lib.LoadAutomationEventList,
	LoadBMFont = lib.LoadBMFont,
	LoadBoneInfoGLTF = lib.LoadBoneInfoGLTF,
	LoadCodepoints = lib.LoadCodepoints,
	LoadDirectoryFiles = lib.LoadDirectoryFiles,
	LoadDirectoryFilesEx = lib.LoadDirectoryFilesEx,
	LoadDroppedFiles = lib.LoadDroppedFiles,
	LoadFileData = lib.LoadFileData,
	LoadFileGLTFCallback = lib.LoadFileGLTFCallback,
	LoadFileText = lib.LoadFileText,
	LoadFont = lib.LoadFont,
	LoadFontData = lib.LoadFontData,
	LoadFontDefault = lib.LoadFontDefault,
	LoadFontEx = lib.LoadFontEx,
	LoadFontFromImage = lib.LoadFontFromImage,
	LoadFontFromMemory = lib.LoadFontFromMemory,
	LoadIQM = lib.LoadIQM,
	LoadImage = lib.LoadImage,
	LoadImageAnim = lib.LoadImageAnim,
	LoadImageAnimFromMemory = lib.LoadImageAnimFromMemory,
	LoadImageColors = lib.LoadImageColors,
	LoadImageFromCgltfImage = lib.LoadImageFromCgltfImage,
	LoadImageFromMemory = lib.LoadImageFromMemory,
	LoadImageFromScreen = lib.LoadImageFromScreen,
	LoadImageFromTexture = lib.LoadImageFromTexture,
	LoadImagePalette = lib.LoadImagePalette,
	LoadImageRaw = lib.LoadImageRaw,
	LoadMaterialDefault = lib.LoadMaterialDefault,
	LoadMaterials = lib.LoadMaterials,
	LoadModel = lib.LoadModel,
	LoadModelAnimations = lib.LoadModelAnimations,
	LoadModelFromMesh = lib.LoadModelFromMesh,
	LoadMusicStream = lib.LoadMusicStream,
	LoadMusicStreamFromMemory = lib.LoadMusicStreamFromMemory,
	LoadOBJ = lib.LoadOBJ,
	LoadRandomSequence = lib.LoadRandomSequence,
	LoadRenderTexture = lib.LoadRenderTexture,
	LoadShader = lib.LoadShader,
	LoadShaderFromMemory = lib.LoadShaderFromMemory,
	LoadSound = lib.LoadSound,
	LoadSoundAlias = lib.LoadSoundAlias,
	LoadSoundFromWave = lib.LoadSoundFromWave,
	LoadTexture = lib.LoadTexture,
	LoadTextureCubemap = lib.LoadTextureCubemap,
	LoadTextureFromImage = lib.LoadTextureFromImage,
	LoadUTF8 = lib.LoadUTF8,
	LoadVrStereoConfig = lib.LoadVrStereoConfig,
	LoadWave = lib.LoadWave,
	LoadWaveFromMemory = lib.LoadWaveFromMemory,
	LoadWaveSamples = lib.LoadWaveSamples,
	MakeDirectory = lib.MakeDirectory,
	MatrixAdd = lib.MatrixAdd,
	MatrixDecompose = lib.MatrixDecompose,
	MatrixDeterminant = lib.MatrixDeterminant,
	MatrixFrustum = lib.MatrixFrustum,
	MatrixIdentity = lib.MatrixIdentity,
	MatrixInvert = lib.MatrixInvert,
	MatrixLookAt = lib.MatrixLookAt,
	MatrixMultiply = lib.MatrixMultiply,
	MatrixOrtho = lib.MatrixOrtho,
	MatrixPerspective = lib.MatrixPerspective,
	MatrixRotate = lib.MatrixRotate,
	MatrixRotateX = lib.MatrixRotateX,
	MatrixRotateXYZ = lib.MatrixRotateXYZ,
	MatrixRotateY = lib.MatrixRotateY,
	MatrixRotateZ = lib.MatrixRotateZ,
	MatrixRotateZYX = lib.MatrixRotateZYX,
	MatrixScale = lib.MatrixScale,
	MatrixSubtract = lib.MatrixSubtract,
	MatrixToFloatV = lib.MatrixToFloatV,
	MatrixTrace = lib.MatrixTrace,
	MatrixTranslate = lib.MatrixTranslate,
	MatrixTranspose = lib.MatrixTranspose,
	MaximizeWindow = lib.MaximizeWindow,
	MeasureText = lib.MeasureText,
	MeasureTextEx = lib.MeasureTextEx,
	MemAlloc = lib.MemAlloc,
	MemFree = lib.MemFree,
	MemRealloc = lib.MemRealloc,
	MinimizeWindow = lib.MinimizeWindow,
	MouseButtonCallback = lib.MouseButtonCallback,
	MouseCursorPosCallback = lib.MouseCursorPosCallback,
	MouseScrollCallback = lib.MouseScrollCallback,
	Normalize = lib.Normalize,
	OnSendAudioDataToDevice = lib.OnSendAudioDataToDevice,
	OpenURL = lib.OpenURL,
	PauseAudioBuffer = lib.PauseAudioBuffer,
	PauseAudioStream = lib.PauseAudioStream,
	PauseMusicStream = lib.PauseMusicStream,
	PauseSound = lib.PauseSound,
	PlayAudioBuffer = lib.PlayAudioBuffer,
	PlayAudioStream = lib.PlayAudioStream,
	PlayAutomationEvent = lib.PlayAutomationEvent,
	PlayMusicStream = lib.PlayMusicStream,
	PlaySound = lib.PlaySound,
	PollInputEvents = lib.PollInputEvents,
	ProcessGestureEvent = lib.ProcessGestureEvent,
	ProcessMaterialsOBJ = lib.ProcessMaterialsOBJ,
	QuaternionAdd = lib.QuaternionAdd,
	QuaternionAddValue = lib.QuaternionAddValue,
	QuaternionCubicHermiteSpline = lib.QuaternionCubicHermiteSpline,
	QuaternionDivide = lib.QuaternionDivide,
	QuaternionEquals = lib.QuaternionEquals,
	QuaternionFromAxisAngle = lib.QuaternionFromAxisAngle,
	QuaternionFromEuler = lib.QuaternionFromEuler,
	QuaternionFromMatrix = lib.QuaternionFromMatrix,
	QuaternionFromVector3ToVector3 = lib.QuaternionFromVector3ToVector3,
	QuaternionIdentity = lib.QuaternionIdentity,
	QuaternionInvert = lib.QuaternionInvert,
	QuaternionLength = lib.QuaternionLength,
	QuaternionLerp = lib.QuaternionLerp,
	QuaternionMultiply = lib.QuaternionMultiply,
	QuaternionNlerp = lib.QuaternionNlerp,
	QuaternionNormalize = lib.QuaternionNormalize,
	QuaternionScale = lib.QuaternionScale,
	QuaternionSlerp = lib.QuaternionSlerp,
	QuaternionSubtract = lib.QuaternionSubtract,
	QuaternionSubtractValue = lib.QuaternionSubtractValue,
	QuaternionToAxisAngle = lib.QuaternionToAxisAngle,
	QuaternionToEuler = lib.QuaternionToEuler,
	QuaternionToMatrix = lib.QuaternionToMatrix,
	QuaternionTransform = lib.QuaternionTransform,
	ReleaseFileGLTFCallback = lib.ReleaseFileGLTFCallback,
	Remap = lib.Remap,
	RestoreWindow = lib.RestoreWindow,
	ResumeAudioBuffer = lib.ResumeAudioBuffer,
	ResumeAudioStream = lib.ResumeAudioStream,
	ResumeMusicStream = lib.ResumeMusicStream,
	ResumeSound = lib.ResumeSound,
	SaveFileData = lib.SaveFileData,
	SaveFileText = lib.SaveFileText,
	ScanDirectoryFiles = lib.ScanDirectoryFiles,
	ScanDirectoryFilesRecursively = lib.ScanDirectoryFilesRecursively,
	SeekMusicStream = lib.SeekMusicStream,
	SetAudioBufferPan = lib.SetAudioBufferPan,
	SetAudioBufferPitch = lib.SetAudioBufferPitch,
	SetAudioBufferVolume = lib.SetAudioBufferVolume,
	SetAudioStreamBufferSizeDefault = lib.SetAudioStreamBufferSizeDefault,
	SetAudioStreamCallback = lib.SetAudioStreamCallback,
	SetAudioStreamPan = lib.SetAudioStreamPan,
	SetAudioStreamPitch = lib.SetAudioStreamPitch,
	SetAudioStreamVolume = lib.SetAudioStreamVolume,
	SetAutomationEventBaseFrame = lib.SetAutomationEventBaseFrame,
	SetAutomationEventList = lib.SetAutomationEventList,
	SetClipboardText = lib.SetClipboardText,
	SetConfigFlags = lib.SetConfigFlags,
	SetDimensionsFromMonitor = lib.SetDimensionsFromMonitor,
	SetExitKey = lib.SetExitKey,
	SetGamepadMappings = lib.SetGamepadMappings,
	SetGamepadVibration = lib.SetGamepadVibration,
	SetGesturesEnabled = lib.SetGesturesEnabled,
	SetLoadFileDataCallback = lib.SetLoadFileDataCallback,
	SetLoadFileTextCallback = lib.SetLoadFileTextCallback,
	SetMasterVolume = lib.SetMasterVolume,
	SetMaterialTexture = lib.SetMaterialTexture,
	SetModelMeshMaterial = lib.SetModelMeshMaterial,
	SetMouseCursor = lib.SetMouseCursor,
	SetMouseOffset = lib.SetMouseOffset,
	SetMousePosition = lib.SetMousePosition,
	SetMouseScale = lib.SetMouseScale,
	SetMusicPan = lib.SetMusicPan,
	SetMusicPitch = lib.SetMusicPitch,
	SetMusicVolume = lib.SetMusicVolume,
	SetPixelColor = lib.SetPixelColor,
	SetRandomSeed = lib.SetRandomSeed,
	SetSaveFileDataCallback = lib.SetSaveFileDataCallback,
	SetSaveFileTextCallback = lib.SetSaveFileTextCallback,
	SetShaderValue = lib.SetShaderValue,
	SetShaderValueMatrix = lib.SetShaderValueMatrix,
	SetShaderValueTexture = lib.SetShaderValueTexture,
	SetShaderValueV = lib.SetShaderValueV,
	SetShapesTexture = lib.SetShapesTexture,
	SetSoundPan = lib.SetSoundPan,
	SetSoundPitch = lib.SetSoundPitch,
	SetSoundVolume = lib.SetSoundVolume,
	SetTargetFPS = lib.SetTargetFPS,
	SetTextLineSpacing = lib.SetTextLineSpacing,
	SetTextureFilter = lib.SetTextureFilter,
	SetTextureWrap = lib.SetTextureWrap,
	SetTraceLogCallback = lib.SetTraceLogCallback,
	SetTraceLogLevel = lib.SetTraceLogLevel,
	SetWindowFocused = lib.SetWindowFocused,
	SetWindowIcon = lib.SetWindowIcon,
	SetWindowIcons = lib.SetWindowIcons,
	SetWindowMaxSize = lib.SetWindowMaxSize,
	SetWindowMinSize = lib.SetWindowMinSize,
	SetWindowMonitor = lib.SetWindowMonitor,
	SetWindowOpacity = lib.SetWindowOpacity,
	SetWindowPosition = lib.SetWindowPosition,
	SetWindowSize = lib.SetWindowSize,
	SetWindowState = lib.SetWindowState,
	SetWindowTitle = lib.SetWindowTitle,
	SetupViewport = lib.SetupViewport,
	ShowCursor = lib.ShowCursor,
	StartAutomationEventRecording = lib.StartAutomationEventRecording,
	StopAudioBuffer = lib.StopAudioBuffer,
	StopAudioBufferInLockedState = lib.StopAudioBufferInLockedState,
	StopAudioStream = lib.StopAudioStream,
	StopAutomationEventRecording = lib.StopAutomationEventRecording,
	StopMusicStream = lib.StopMusicStream,
	StopSound = lib.StopSound,
	SwapScreenBuffer = lib.SwapScreenBuffer,
	TakeScreenshot = lib.TakeScreenshot,
	TextAppend = lib.TextAppend,
	TextCopy = lib.TextCopy,
	TextFindIndex = lib.TextFindIndex,
	TextFormat = lib.TextFormat,
	TextInsert = lib.TextInsert,
	TextIsEqual = lib.TextIsEqual,
	TextJoin = lib.TextJoin,
	TextLength = lib.TextLength,
	TextReplace = lib.TextReplace,
	TextSplit = lib.TextSplit,
	TextSubtext = lib.TextSubtext,
	TextToCamel = lib.TextToCamel,
	TextToFloat = lib.TextToFloat,
	TextToInteger = lib.TextToInteger,
	TextToLower = lib.TextToLower,
	TextToPascal = lib.TextToPascal,
	TextToSnake = lib.TextToSnake,
	TextToUpper = lib.TextToUpper,
	ToggleBorderlessWindowed = lib.ToggleBorderlessWindowed,
	ToggleFullscreen = lib.ToggleFullscreen,
	TraceLog = lib.TraceLog,
	TrackAudioBuffer = lib.TrackAudioBuffer,
	UnloadAudioBuffer = lib.UnloadAudioBuffer,
	UnloadAudioStream = lib.UnloadAudioStream,
	UnloadAutomationEventList = lib.UnloadAutomationEventList,
	UnloadCodepoints = lib.UnloadCodepoints,
	UnloadDirectoryFiles = lib.UnloadDirectoryFiles,
	UnloadDroppedFiles = lib.UnloadDroppedFiles,
	UnloadFileData = lib.UnloadFileData,
	UnloadFileText = lib.UnloadFileText,
	UnloadFont = lib.UnloadFont,
	UnloadFontData = lib.UnloadFontData,
	UnloadFontDefault = lib.UnloadFontDefault,
	UnloadImage = lib.UnloadImage,
	UnloadImageColors = lib.UnloadImageColors,
	UnloadImagePalette = lib.UnloadImagePalette,
	UnloadMaterial = lib.UnloadMaterial,
	UnloadMesh = lib.UnloadMesh,
	UnloadModel = lib.UnloadModel,
	UnloadModelAnimation = lib.UnloadModelAnimation,
	UnloadModelAnimations = lib.UnloadModelAnimations,
	UnloadMusicStream = lib.UnloadMusicStream,
	UnloadRandomSequence = lib.UnloadRandomSequence,
	UnloadRenderTexture = lib.UnloadRenderTexture,
	UnloadShader = lib.UnloadShader,
	UnloadSound = lib.UnloadSound,
	UnloadSoundAlias = lib.UnloadSoundAlias,
	UnloadTexture = lib.UnloadTexture,
	UnloadUTF8 = lib.UnloadUTF8,
	UnloadVrStereoConfig = lib.UnloadVrStereoConfig,
	UnloadWave = lib.UnloadWave,
	UnloadWaveSamples = lib.UnloadWaveSamples,
	UntrackAudioBuffer = lib.UntrackAudioBuffer,
	UpdateAudioStream = lib.UpdateAudioStream,
	UpdateAudioStreamInLockedState = lib.UpdateAudioStreamInLockedState,
	UpdateCamera = lib.UpdateCamera,
	UpdateCameraPro = lib.UpdateCameraPro,
	UpdateGestures = lib.UpdateGestures,
	UpdateMeshBuffer = lib.UpdateMeshBuffer,
	UpdateModelAnimation = lib.UpdateModelAnimation,
	UpdateModelAnimationBones = lib.UpdateModelAnimationBones,
	UpdateMusicStream = lib.UpdateMusicStream,
	UpdateSound = lib.UpdateSound,
	UpdateTexture = lib.UpdateTexture,
	UpdateTextureRec = lib.UpdateTextureRec,
	UploadMesh = lib.UploadMesh,
	Vector2Add = lib.Vector2Add,
	Vector2AddValue = lib.Vector2AddValue,
	Vector2Angle = lib.Vector2Angle,
	Vector2Clamp = lib.Vector2Clamp,
	Vector2ClampValue = lib.Vector2ClampValue,
	Vector2Distance = lib.Vector2Distance,
	Vector2DistanceSqr = lib.Vector2DistanceSqr,
	Vector2Divide = lib.Vector2Divide,
	Vector2DotProduct = lib.Vector2DotProduct,
	Vector2Equals = lib.Vector2Equals,
	Vector2Invert = lib.Vector2Invert,
	Vector2Length = lib.Vector2Length,
	Vector2LengthSqr = lib.Vector2LengthSqr,
	Vector2Lerp = lib.Vector2Lerp,
	Vector2LineAngle = lib.Vector2LineAngle,
	Vector2Max = lib.Vector2Max,
	Vector2Min = lib.Vector2Min,
	Vector2MoveTowards = lib.Vector2MoveTowards,
	Vector2Multiply = lib.Vector2Multiply,
	Vector2Negate = lib.Vector2Negate,
	Vector2Normalize = lib.Vector2Normalize,
	Vector2One = lib.Vector2One,
	Vector2Reflect = lib.Vector2Reflect,
	Vector2Refract = lib.Vector2Refract,
	Vector2Rotate = lib.Vector2Rotate,
	Vector2Scale = lib.Vector2Scale,
	Vector2Subtract = lib.Vector2Subtract,
	Vector2SubtractValue = lib.Vector2SubtractValue,
	Vector2Transform = lib.Vector2Transform,
	Vector2Zero = lib.Vector2Zero,
	Vector3Add = lib.Vector3Add,
	Vector3AddValue = lib.Vector3AddValue,
	Vector3Angle = lib.Vector3Angle,
	Vector3Barycenter = lib.Vector3Barycenter,
	Vector3Clamp = lib.Vector3Clamp,
	Vector3ClampValue = lib.Vector3ClampValue,
	Vector3CrossProduct = lib.Vector3CrossProduct,
	Vector3CubicHermite = lib.Vector3CubicHermite,
	Vector3Distance = lib.Vector3Distance,
	Vector3DistanceSqr = lib.Vector3DistanceSqr,
	Vector3Divide = lib.Vector3Divide,
	Vector3DotProduct = lib.Vector3DotProduct,
	Vector3Equals = lib.Vector3Equals,
	Vector3Invert = lib.Vector3Invert,
	Vector3Length = lib.Vector3Length,
	Vector3LengthSqr = lib.Vector3LengthSqr,
	Vector3Lerp = lib.Vector3Lerp,
	Vector3Max = lib.Vector3Max,
	Vector3Min = lib.Vector3Min,
	Vector3MoveTowards = lib.Vector3MoveTowards,
	Vector3Multiply = lib.Vector3Multiply,
	Vector3Negate = lib.Vector3Negate,
	Vector3Normalize = lib.Vector3Normalize,
	Vector3One = lib.Vector3One,
	Vector3OrthoNormalize = lib.Vector3OrthoNormalize,
	Vector3Perpendicular = lib.Vector3Perpendicular,
	Vector3Project = lib.Vector3Project,
	Vector3Reflect = lib.Vector3Reflect,
	Vector3Refract = lib.Vector3Refract,
	Vector3Reject = lib.Vector3Reject,
	Vector3RotateByAxisAngle = lib.Vector3RotateByAxisAngle,
	Vector3RotateByQuaternion = lib.Vector3RotateByQuaternion,
	Vector3Scale = lib.Vector3Scale,
	Vector3Subtract = lib.Vector3Subtract,
	Vector3SubtractValue = lib.Vector3SubtractValue,
	Vector3ToFloatV = lib.Vector3ToFloatV,
	Vector3Transform = lib.Vector3Transform,
	Vector3Unproject = lib.Vector3Unproject,
	Vector3Zero = lib.Vector3Zero,
	Vector4Add = lib.Vector4Add,
	Vector4AddValue = lib.Vector4AddValue,
	Vector4Distance = lib.Vector4Distance,
	Vector4DistanceSqr = lib.Vector4DistanceSqr,
	Vector4Divide = lib.Vector4Divide,
	Vector4DotProduct = lib.Vector4DotProduct,
	Vector4Equals = lib.Vector4Equals,
	Vector4Invert = lib.Vector4Invert,
	Vector4Length = lib.Vector4Length,
	Vector4LengthSqr = lib.Vector4LengthSqr,
	Vector4Lerp = lib.Vector4Lerp,
	Vector4Max = lib.Vector4Max,
	Vector4Min = lib.Vector4Min,
	Vector4MoveTowards = lib.Vector4MoveTowards,
	Vector4Multiply = lib.Vector4Multiply,
	Vector4Negate = lib.Vector4Negate,
	Vector4Normalize = lib.Vector4Normalize,
	Vector4One = lib.Vector4One,
	Vector4Scale = lib.Vector4Scale,
	Vector4Subtract = lib.Vector4Subtract,
	Vector4SubtractValue = lib.Vector4SubtractValue,
	Vector4Zero = lib.Vector4Zero,
	Vox_FreeArrays = lib.Vox_FreeArrays,
	Vox_GetVoxel = lib.Vox_GetVoxel,
	Vox_LoadFromMemory = lib.Vox_LoadFromMemory,
	WaitTime = lib.WaitTime,
	WaveCopy = lib.WaveCopy,
	WaveCrop = lib.WaveCrop,
	WaveFormat = lib.WaveFormat,
	WindowContentScaleCallback = lib.WindowContentScaleCallback,
	WindowDropCallback = lib.WindowDropCallback,
	WindowFocusCallback = lib.WindowFocusCallback,
	WindowIconifyCallback = lib.WindowIconifyCallback,
	WindowMaximizeCallback = lib.WindowMaximizeCallback,
	WindowPosCallback = lib.WindowPosCallback,
	WindowShouldClose = lib.WindowShouldClose,
	WindowSizeCallback = lib.WindowSizeCallback,
	Wrap = lib.Wrap,

	rlDisableDepthTest = lib.rlDisableDepthTest,
	rlEnableDepthTest = lib.rlEnableDepthTest,
	rlDisableDepthMask = lib.rlDisableDepthMask,
	rlEnableDepthMask = lib.rlEnableDepthMask,
	rlDisableBackfaceCulling = lib.rlDisableBackfaceCulling,
	rlEnableBackfaceCulling = lib.rlEnableBackfaceCulling,
	rlSetClipPlanes = lib.rlSetClipPlanes,
	rlGetCullDistanceNear = lib.rlGetCullDistanceNear,
	rlGetCullDistanceFar = lib.rlGetCullDistanceFar,

	GuiWindowBox = lib.GuiWindowBox,
	lib = lib,

	-- rip not organized
	-- well i used luau to get this list quickly so good enough

	CAMERA_ORTHOGRAPHIC = const.CAMERA_ORTHOGRAPHIC,
	MOUSE_BUTTON_SIDE = const.MOUSE_BUTTON_SIDE,
	KEY_A = const.KEY_A,
	KEY_KP_3 = const.KEY_KP_3,
	PIXELFORMAT_UNCOMPRESSED_R16G16B16 = const.PIXELFORMAT_UNCOMPRESSED_R16G16B16,
	KEY_TWO = const.KEY_TWO,
	SHADER_LOC_COLOR_SPECULAR = const.SHADER_LOC_COLOR_SPECULAR,
	Mesh = const.Mesh,
	LIGHTGRAY = const.LIGHTGRAY,
	DARKGREEN = const.DARKGREEN,
	BLEND_ALPHA_PREMULTIPLY = const.BLEND_ALPHA_PREMULTIPLY,
	KEY_KP_DECIMAL = const.KEY_KP_DECIMAL,
	MOUSE_BUTTON_MIDDLE = const.MOUSE_BUTTON_MIDDLE,
	PIXELFORMAT_COMPRESSED_ETC2_EAC_RGBA = const.PIXELFORMAT_COMPRESSED_ETC2_EAC_RGBA,
	Transform = const.Transform,
	GAMEPAD_BUTTON_UNKNOWN = const.GAMEPAD_BUTTON_UNKNOWN,
	LOG_FATAL = const.LOG_FATAL,
	SHADER_LOC_MATRIX_NORMAL = const.SHADER_LOC_MATRIX_NORMAL,
	BLEND_MULTIPLIED = const.BLEND_MULTIPLIED,
	KEY_LEFT_ALT = const.KEY_LEFT_ALT,
	KEY_RIGHT_SUPER = const.KEY_RIGHT_SUPER,
	KEY_TAB = const.KEY_TAB,
	KEY_KP_5 = const.KEY_KP_5,
	GAMEPAD_BUTTON_RIGHT_FACE_RIGHT = const.GAMEPAD_BUTTON_RIGHT_FACE_RIGHT,
	KEY_D = const.KEY_D,
	BLANK = const.BLANK,
	Font = const.Font,
	MATERIAL_MAP_IRRADIANCE = const.MATERIAL_MAP_IRRADIANCE,
	SHADER_LOC_MAP_CUBEMAP = const.SHADER_LOC_MAP_CUBEMAP,
	KEY_O = const.KEY_O,
	MaterialMap = const.MaterialMap,
	SHADER_UNIFORM_INT = const.SHADER_UNIFORM_INT,
	KEY_PRINT_SCREEN = const.KEY_PRINT_SCREEN,
	FLAG_WINDOW_TRANSPARENT = const.FLAG_WINDOW_TRANSPARENT,
	GAMEPAD_BUTTON_RIGHT_TRIGGER_1 = const.GAMEPAD_BUTTON_RIGHT_TRIGGER_1,
	KEY_C = const.KEY_C,
	FLAG_WINDOW_HIGHDPI = const.FLAG_WINDOW_HIGHDPI,
	Camera2D = const.Camera2D,
	SHADER_LOC_COLOR_DIFFUSE = const.SHADER_LOC_COLOR_DIFFUSE,
	Texture = const.Texture,
	FLAG_WINDOW_MAXIMIZED = const.FLAG_WINDOW_MAXIMIZED,
	KEY_G = const.KEY_G,
	GESTURE_TAP = const.GESTURE_TAP,
	PIXELFORMAT_COMPRESSED_DXT3_RGBA = const.PIXELFORMAT_COMPRESSED_DXT3_RGBA,
	MATERIAL_MAP_SPECULAR = const.MATERIAL_MAP_SPECULAR,
	SHADER_UNIFORM_VEC3 = const.SHADER_UNIFORM_VEC3,
	Quaternion = const.Quaternion,
	CAMERA_FIRST_PERSON = const.CAMERA_FIRST_PERSON,
	FLAG_WINDOW_RESIZABLE = const.FLAG_WINDOW_RESIZABLE,
	GOLD = const.GOLD,
	FLAG_WINDOW_TOPMOST = const.FLAG_WINDOW_TOPMOST,
	AutomationEventList = const.AutomationEventList,
	VrStereoConfig = const.VrStereoConfig,
	KEY_P = const.KEY_P,
	rAudioProcessor = const.rAudioProcessor,
	KEY_GRAVE = const.KEY_GRAVE,
	SHADER_LOC_MATRIX_MVP = const.SHADER_LOC_MATRIX_MVP,
	MATERIAL_MAP_DIFFUSE = const.MATERIAL_MAP_DIFFUSE,
	KEY_KP_ADD = const.KEY_KP_ADD,
	SHADER_UNIFORM_IVEC2 = const.SHADER_UNIFORM_IVEC2,
	Wave = const.Wave,
	KEY_KP_4 = const.KEY_KP_4,
	RenderTexture = const.RenderTexture,
	PIXELFORMAT_COMPRESSED_ETC1_RGB = const.PIXELFORMAT_COMPRESSED_ETC1_RGB,
	SHADER_LOC_MATRIX_VIEW = const.SHADER_LOC_MATRIX_VIEW,
	KEY_PAGE_DOWN = const.KEY_PAGE_DOWN,
	MOUSE_CURSOR_RESIZE_NESW = const.MOUSE_CURSOR_RESIZE_NESW,
	KEY_KP_2 = const.KEY_KP_2,
	FLAG_WINDOW_UNDECORATED = const.FLAG_WINDOW_UNDECORATED,
	LOG_TRACE = const.LOG_TRACE,
	KEY_RIGHT_CONTROL = const.KEY_RIGHT_CONTROL,
	LIME = const.LIME,
	GlyphInfo = const.GlyphInfo,
	KEY_DELETE = const.KEY_DELETE,
	PIXELFORMAT_UNCOMPRESSED_R8G8B8A8 = const.PIXELFORMAT_UNCOMPRESSED_R8G8B8A8,
	KEY_SCROLL_LOCK = const.KEY_SCROLL_LOCK,
	KEY_F3 = const.KEY_F3,
	MATERIAL_MAP_BRDF = const.MATERIAL_MAP_BRDF,
	SHADER_LOC_COLOR_AMBIENT = const.SHADER_LOC_COLOR_AMBIENT,
	MATERIAL_MAP_METALNESS = const.MATERIAL_MAP_METALNESS,
	RenderTexture2D = const.RenderTexture2D,
	KEY_CAPS_LOCK = const.KEY_CAPS_LOCK,
	CAMERA_PERSPECTIVE = const.CAMERA_PERSPECTIVE,
	GAMEPAD_AXIS_RIGHT_X = const.GAMEPAD_AXIS_RIGHT_X,
	Color = const.Color,
	DARKPURPLE = const.DARKPURPLE,
	KEY_F11 = const.KEY_F11,
	MATERIAL_MAP_OCCLUSION = const.MATERIAL_MAP_OCCLUSION,
	KEY_MINUS = const.KEY_MINUS,
	PIXELFORMAT_COMPRESSED_PVRT_RGB = const.PIXELFORMAT_COMPRESSED_PVRT_RGB,
	SHADER_LOC_VERTEX_TANGENT = const.SHADER_LOC_VERTEX_TANGENT,
	TEXTURE_FILTER_TRILINEAR = const.TEXTURE_FILTER_TRILINEAR,
	KEY_F12 = const.KEY_F12,
	TEXTURE_WRAP_REPEAT = const.TEXTURE_WRAP_REPEAT,
	PINK = const.PINK,
	KEY_K = const.KEY_K,
	KEY_FOUR = const.KEY_FOUR,
	KEY_PAGE_UP = const.KEY_PAGE_UP,
	SHADER_UNIFORM_IVEC3 = const.SHADER_UNIFORM_IVEC3,
	GREEN = const.GREEN,
	LOG_NONE = const.LOG_NONE,
	GRAY = const.GRAY,
	KEY_KP_8 = const.KEY_KP_8,
	PIXELFORMAT_COMPRESSED_ASTC_4x4_RGBA = const.PIXELFORMAT_COMPRESSED_ASTC_4x4_RGBA,
	PIXELFORMAT_UNCOMPRESSED_R32 = const.PIXELFORMAT_UNCOMPRESSED_R32,
	KEY_PERIOD = const.KEY_PERIOD,
	AutomationEvent = const.AutomationEvent,
	LOG_INFO = const.LOG_INFO,
	SHADER_LOC_MAP_PREFILTER = const.SHADER_LOC_MAP_PREFILTER,
	KEY_VOLUME_DOWN = const.KEY_VOLUME_DOWN,
	KEY_Z = const.KEY_Z,
	GESTURE_NONE = const.GESTURE_NONE,
	GAMEPAD_BUTTON_LEFT_TRIGGER_2 = const.GAMEPAD_BUTTON_LEFT_TRIGGER_2,
	MATERIAL_MAP_EMISSION = const.MATERIAL_MAP_EMISSION,
	FLAG_WINDOW_MINIMIZED = const.FLAG_WINDOW_MINIMIZED,
	KEY_NINE = const.KEY_NINE,
	KEY_T = const.KEY_T,
	NPATCH_THREE_PATCH_HORIZONTAL = const.NPATCH_THREE_PATCH_HORIZONTAL,
	FLAG_WINDOW_ALWAYS_RUN = const.FLAG_WINDOW_ALWAYS_RUN,
	SHADER_LOC_VERTEX_COLOR = const.SHADER_LOC_VERTEX_COLOR,
	KEY_RIGHT_SHIFT = const.KEY_RIGHT_SHIFT,
	FLAG_FULLSCREEN_MODE = const.FLAG_FULLSCREEN_MODE,
	KEY_X = const.KEY_X,
	RayCollision = const.RayCollision,
	rAudioBuffer = const.rAudioBuffer,
	MOUSE_CURSOR_POINTING_HAND = const.MOUSE_CURSOR_POINTING_HAND,
	KEY_J = const.KEY_J,
	Matrix = const.Matrix,
	TEXTURE_FILTER_ANISOTROPIC_8X = const.TEXTURE_FILTER_ANISOTROPIC_8X,
	MOUSE_BUTTON_RIGHT = const.MOUSE_BUTTON_RIGHT,
	KEY_S = const.KEY_S,
	SKYBLUE = const.SKYBLUE,
	KEY_BACKSLASH = const.KEY_BACKSLASH,
	PIXELFORMAT_UNCOMPRESSED_R8G8B8 = const.PIXELFORMAT_UNCOMPRESSED_R8G8B8,
	KEY_LEFT = const.KEY_LEFT,
	RED = const.RED,
	GESTURE_HOLD = const.GESTURE_HOLD,
	VIOLET = const.VIOLET,
	KEY_SEMICOLON = const.KEY_SEMICOLON,
	FLAG_MSAA_4X_HINT = const.FLAG_MSAA_4X_HINT,
	BLEND_CUSTOM_SEPARATE = const.BLEND_CUSTOM_SEPARATE,
	KEY_PAUSE = const.KEY_PAUSE,
	GAMEPAD_BUTTON_RIGHT_FACE_DOWN = const.GAMEPAD_BUTTON_RIGHT_FACE_DOWN,
	NPatchInfo = const.NPatchInfo,
	KEY_F1 = const.KEY_F1,
	SHADER_LOC_MAP_DIFFUSE = const.SHADER_LOC_MAP_DIFFUSE,
	DARKBLUE = const.DARKBLUE,
	NPATCH_THREE_PATCH_VERTICAL = const.NPATCH_THREE_PATCH_VERTICAL,
	NPATCH_NINE_PATCH = const.NPATCH_NINE_PATCH,
	GAMEPAD_AXIS_RIGHT_TRIGGER = const.GAMEPAD_AXIS_RIGHT_TRIGGER,
	KEY_ZERO = const.KEY_ZERO,
	TEXTURE_WRAP_MIRROR_REPEAT = const.TEXTURE_WRAP_MIRROR_REPEAT,
	CAMERA_ORBITAL = const.CAMERA_ORBITAL,
	CAMERA_FREE = const.CAMERA_FREE,
	FLAG_WINDOW_MOUSE_PASSTHROUGH = const.FLAG_WINDOW_MOUSE_PASSTHROUGH,
	CAMERA_CUSTOM = const.CAMERA_CUSTOM,
	GESTURE_DOUBLETAP = const.GESTURE_DOUBLETAP,
	GESTURE_PINCH_OUT = const.GESTURE_PINCH_OUT,
	GESTURE_PINCH_IN = const.GESTURE_PINCH_IN,
	KEY_SPACE = const.KEY_SPACE,
	BLEND_SUBTRACT_COLORS = const.BLEND_SUBTRACT_COLORS,
	Material = const.Material,
	MOUSE_CURSOR_RESIZE_ALL = const.MOUSE_CURSOR_RESIZE_ALL,
	GESTURE_SWIPE_LEFT = const.GESTURE_SWIPE_LEFT,
	GAMEPAD_AXIS_LEFT_X = const.GAMEPAD_AXIS_LEFT_X,
	GESTURE_SWIPE_RIGHT = const.GESTURE_SWIPE_RIGHT,
	KEY_H = const.KEY_H,
	LOG_DEBUG = const.LOG_DEBUG,
	BLEND_CUSTOM = const.BLEND_CUSTOM,
	GESTURE_SWIPE_UP = const.GESTURE_SWIPE_UP,
	PIXELFORMAT_COMPRESSED_ETC2_RGB = const.PIXELFORMAT_COMPRESSED_ETC2_RGB,
	KEY_INSERT = const.KEY_INSERT,
	WHITE = const.WHITE,
	BLEND_ALPHA = const.BLEND_ALPHA,
	KEY_R = const.KEY_R,
	Model = const.Model,
	KEY_KP_ENTER = const.KEY_KP_ENTER,
	FONT_SDF = const.FONT_SDF,
	FONT_BITMAP = const.FONT_BITMAP,
	FONT_DEFAULT = const.FONT_DEFAULT,
	KEY_KP_9 = const.KEY_KP_9,
	CUBEMAP_LAYOUT_CROSS_FOUR_BY_THREE = const.CUBEMAP_LAYOUT_CROSS_FOUR_BY_THREE,
	CUBEMAP_LAYOUT_CROSS_THREE_BY_FOUR = const.CUBEMAP_LAYOUT_CROSS_THREE_BY_FOUR,
	CUBEMAP_LAYOUT_LINE_HORIZONTAL = const.CUBEMAP_LAYOUT_LINE_HORIZONTAL,
	MOUSE_CURSOR_ARROW = const.MOUSE_CURSOR_ARROW,
	SHADER_LOC_MAP_ALBEDO = const.SHADER_LOC_MAP_ALBEDO,
	GAMEPAD_BUTTON_RIGHT_FACE_UP = const.GAMEPAD_BUTTON_RIGHT_FACE_UP,
	Rectangle = const.Rectangle,
	KEY_UP = const.KEY_UP,
	GAMEPAD_BUTTON_LEFT_FACE_DOWN = const.GAMEPAD_BUTTON_LEFT_FACE_DOWN,
	KEY_F2 = const.KEY_F2,
	GESTURE_DRAG = const.GESTURE_DRAG,
	TEXTURE_WRAP_MIRROR_CLAMP = const.TEXTURE_WRAP_MIRROR_CLAMP,
	KEY_RIGHT_ALT = const.KEY_RIGHT_ALT,
	TEXTURE_WRAP_CLAMP = const.TEXTURE_WRAP_CLAMP,
	Vector3 = const.Vector3,
	TEXTURE_FILTER_ANISOTROPIC_16X = const.TEXTURE_FILTER_ANISOTROPIC_16X,
	TEXTURE_FILTER_ANISOTROPIC_4X = const.TEXTURE_FILTER_ANISOTROPIC_4X,
	KEY_KP_MULTIPLY = const.KEY_KP_MULTIPLY,
	TEXTURE_FILTER_BILINEAR = const.TEXTURE_FILTER_BILINEAR,
	PIXELFORMAT_UNCOMPRESSED_R4G4B4A4 = const.PIXELFORMAT_UNCOMPRESSED_R4G4B4A4,
	TEXTURE_FILTER_POINT = const.TEXTURE_FILTER_POINT,
	SHADER_LOC_VERTEX_NORMAL = const.SHADER_LOC_VERTEX_NORMAL,
	KEY_L = const.KEY_L,
	PIXELFORMAT_COMPRESSED_ASTC_8x8_RGBA = const.PIXELFORMAT_COMPRESSED_ASTC_8x8_RGBA,
	Camera = const.Camera,
	PURPLE = const.PURPLE,
	PIXELFORMAT_COMPRESSED_PVRT_RGBA = const.PIXELFORMAT_COMPRESSED_PVRT_RGBA,
	BLEND_ADD_COLORS = const.BLEND_ADD_COLORS,
	MAGENTA = const.MAGENTA,
	PIXELFORMAT_COMPRESSED_DXT5_RGBA = const.PIXELFORMAT_COMPRESSED_DXT5_RGBA,
	PIXELFORMAT_UNCOMPRESSED_R16G16B16A16 = const.PIXELFORMAT_UNCOMPRESSED_R16G16B16A16,
	MOUSE_CURSOR_NOT_ALLOWED = const.MOUSE_CURSOR_NOT_ALLOWED,
	PIXELFORMAT_UNCOMPRESSED_R16 = const.PIXELFORMAT_UNCOMPRESSED_R16,
	SHADER_UNIFORM_VEC4 = const.SHADER_UNIFORM_VEC4,
	PIXELFORMAT_COMPRESSED_DXT1_RGB = const.PIXELFORMAT_COMPRESSED_DXT1_RGB,
	PIXELFORMAT_COMPRESSED_DXT1_RGBA = const.PIXELFORMAT_COMPRESSED_DXT1_RGBA,
	FLAG_WINDOW_UNFOCUSED = const.FLAG_WINDOW_UNFOCUSED,
	PIXELFORMAT_UNCOMPRESSED_R32G32B32A32 = const.PIXELFORMAT_UNCOMPRESSED_R32G32B32A32,
	SHADER_LOC_MAP_IRRADIANCE = const.SHADER_LOC_MAP_IRRADIANCE,
	PIXELFORMAT_UNCOMPRESSED_R32G32B32 = const.PIXELFORMAT_UNCOMPRESSED_R32G32B32,
	PIXELFORMAT_UNCOMPRESSED_R5G5B5A1 = const.PIXELFORMAT_UNCOMPRESSED_R5G5B5A1,
	KEY_HOME = const.KEY_HOME,
	SHADER_LOC_MAP_HEIGHT = const.SHADER_LOC_MAP_HEIGHT,
	PIXELFORMAT_UNCOMPRESSED_R5G6B5 = const.PIXELFORMAT_UNCOMPRESSED_R5G6B5,
	VrDeviceInfo = const.VrDeviceInfo,
	Sound = const.Sound,
	BoundingBox = const.BoundingBox,
	FLAG_WINDOW_HIDDEN = const.FLAG_WINDOW_HIDDEN,
	YELLOW = const.YELLOW,
	GAMEPAD_BUTTON_RIGHT_FACE_LEFT = const.GAMEPAD_BUTTON_RIGHT_FACE_LEFT,
	LOG_ERROR = const.LOG_ERROR,
	KEY_EIGHT = const.KEY_EIGHT,
	KEY_RIGHT = const.KEY_RIGHT,
	GAMEPAD_BUTTON_MIDDLE_LEFT = const.GAMEPAD_BUTTON_MIDDLE_LEFT,
	MATERIAL_MAP_PREFILTER = const.MATERIAL_MAP_PREFILTER,
	SHADER_LOC_MAP_OCCLUSION = const.SHADER_LOC_MAP_OCCLUSION,
	SHADER_LOC_MAP_SPECULAR = const.SHADER_LOC_MAP_SPECULAR,
	SHADER_ATTRIB_VEC2 = const.SHADER_ATTRIB_VEC2,
	KEY_SEVEN = const.KEY_SEVEN,
	SHADER_ATTRIB_FLOAT = const.SHADER_ATTRIB_FLOAT,
	Music = const.Music,
	SHADER_UNIFORM_SAMPLER2D = const.SHADER_UNIFORM_SAMPLER2D,
	KEY_KP_0 = const.KEY_KP_0,
	SHADER_UNIFORM_IVEC4 = const.SHADER_UNIFORM_IVEC4,
	SHADER_UNIFORM_VEC2 = const.SHADER_UNIFORM_VEC2,
	KEY_KP_DIVIDE = const.KEY_KP_DIVIDE,
	SHADER_UNIFORM_FLOAT = const.SHADER_UNIFORM_FLOAT,
	MOUSE_BUTTON_BACK = const.MOUSE_BUTTON_BACK,
	SHADER_LOC_BONE_MATRICES = const.SHADER_LOC_BONE_MATRICES,
	SHADER_LOC_VERTEX_BONEWEIGHTS = const.SHADER_LOC_VERTEX_BONEWEIGHTS,
	KEY_E = const.KEY_E,
	SHADER_LOC_VERTEX_BONEIDS = const.SHADER_LOC_VERTEX_BONEIDS,
	BoneInfo = const.BoneInfo,
	SHADER_LOC_MAP_BRDF = const.SHADER_LOC_MAP_BRDF,
	SHADER_LOC_MAP_EMISSION = const.SHADER_LOC_MAP_EMISSION,
	SHADER_LOC_MAP_ROUGHNESS = const.SHADER_LOC_MAP_ROUGHNESS,
	KEY_FIVE = const.KEY_FIVE,
	GAMEPAD_BUTTON_LEFT_FACE_LEFT = const.GAMEPAD_BUTTON_LEFT_FACE_LEFT,
	KEY_LEFT_SHIFT = const.KEY_LEFT_SHIFT,
	KEY_B = const.KEY_B,
	SHADER_LOC_MAP_METALNESS = const.SHADER_LOC_MAP_METALNESS,
	KEY_KP_6 = const.KEY_KP_6,
	CUBEMAP_LAYOUT_LINE_VERTICAL = const.CUBEMAP_LAYOUT_LINE_VERTICAL,
	Camera3D = const.Camera3D,
	SHADER_LOC_VECTOR_VIEW = const.SHADER_LOC_VECTOR_VIEW,
	BROWN = const.BROWN,
	SHADER_LOC_MATRIX_MODEL = const.SHADER_LOC_MATRIX_MODEL,
	SHADER_LOC_MATRIX_PROJECTION = const.SHADER_LOC_MATRIX_PROJECTION,
	Texture2D = const.Texture2D,
	MOUSE_CURSOR_IBEAM = const.MOUSE_CURSOR_IBEAM,
	KEY_SIX = const.KEY_SIX,
	SHADER_LOC_VERTEX_TEXCOORD02 = const.SHADER_LOC_VERTEX_TEXCOORD02,
	SHADER_LOC_VERTEX_TEXCOORD01 = const.SHADER_LOC_VERTEX_TEXCOORD01,
	KEY_MENU = const.KEY_MENU,
	SHADER_LOC_VERTEX_POSITION = const.SHADER_LOC_VERTEX_POSITION,
	SHADER_ATTRIB_VEC3 = const.SHADER_ATTRIB_VEC3,
	FilePathList = const.FilePathList,
	MATERIAL_MAP_CUBEMAP = const.MATERIAL_MAP_CUBEMAP,
	KEY_THREE = const.KEY_THREE,
	MATERIAL_MAP_HEIGHT = const.MATERIAL_MAP_HEIGHT,
	MOUSE_BUTTON_FORWARD = const.MOUSE_BUTTON_FORWARD,
	KEY_BACKSPACE = const.KEY_BACKSPACE,
	KEY_U = const.KEY_U,
	MATERIAL_MAP_NORMAL = const.MATERIAL_MAP_NORMAL,
	Image = const.Image,
	DARKBROWN = const.DARKBROWN,
	MATERIAL_MAP_ALBEDO = const.MATERIAL_MAP_ALBEDO,
	KEY_Y = const.KEY_Y,
	CAMERA_THIRD_PERSON = const.CAMERA_THIRD_PERSON,
	BLACK = const.BLACK,
	Vector4 = const.Vector4,
	AudioStream = const.AudioStream,
	GAMEPAD_AXIS_RIGHT_Y = const.GAMEPAD_AXIS_RIGHT_Y,
	MOUSE_BUTTON_EXTRA = const.MOUSE_BUTTON_EXTRA,
	GAMEPAD_BUTTON_RIGHT_THUMB = const.GAMEPAD_BUTTON_RIGHT_THUMB,
	GAMEPAD_BUTTON_LEFT_THUMB = const.GAMEPAD_BUTTON_LEFT_THUMB,
	GAMEPAD_BUTTON_MIDDLE_RIGHT = const.GAMEPAD_BUTTON_MIDDLE_RIGHT,
	GAMEPAD_BUTTON_MIDDLE = const.GAMEPAD_BUTTON_MIDDLE,
	SHADER_LOC_MAP_NORMAL = const.SHADER_LOC_MAP_NORMAL,
	Ray = const.Ray,
	GAMEPAD_BUTTON_LEFT_TRIGGER_1 = const.GAMEPAD_BUTTON_LEFT_TRIGGER_1,
	PIXELFORMAT_UNCOMPRESSED_GRAY_ALPHA = const.PIXELFORMAT_UNCOMPRESSED_GRAY_ALPHA,
	Vector2 = const.Vector2,
	KEY_BACK = const.KEY_BACK,
	GAMEPAD_BUTTON_LEFT_FACE_RIGHT = const.GAMEPAD_BUTTON_LEFT_FACE_RIGHT,
	GAMEPAD_BUTTON_LEFT_FACE_UP = const.GAMEPAD_BUTTON_LEFT_FACE_UP,
	KEY_F8 = const.KEY_F8,
	MOUSE_CURSOR_RESIZE_NWSE = const.MOUSE_CURSOR_RESIZE_NWSE,
	KEY_F6 = const.KEY_F6,
	LOG_WARNING = const.LOG_WARNING,
	MOUSE_CURSOR_RESIZE_NS = const.MOUSE_CURSOR_RESIZE_NS,
	MOUSE_CURSOR_RESIZE_EW = const.MOUSE_CURSOR_RESIZE_EW,
	FLAG_BORDERLESS_WINDOWED_MODE = const.FLAG_BORDERLESS_WINDOWED_MODE,
	MOUSE_CURSOR_CROSSHAIR = const.MOUSE_CURSOR_CROSSHAIR,
	KEY_LEFT_SUPER = const.KEY_LEFT_SUPER,
	KEY_LEFT_BRACKET = const.KEY_LEFT_BRACKET,
	FLAG_VSYNC_HINT = const.FLAG_VSYNC_HINT,
	MATERIAL_MAP_ROUGHNESS = const.MATERIAL_MAP_ROUGHNESS,
	KEY_F4 = const.KEY_F4,
	TextureCubemap = const.TextureCubemap,
	GAMEPAD_AXIS_LEFT_Y = const.GAMEPAD_AXIS_LEFT_Y,
	MOUSE_BUTTON_LEFT = const.MOUSE_BUTTON_LEFT,
	KEY_VOLUME_UP = const.KEY_VOLUME_UP,
	KEY_KP_EQUAL = const.KEY_KP_EQUAL,
	RAYWHITE = const.RAYWHITE,
	DARKGRAY = const.DARKGRAY,
	KEY_NULL = const.KEY_NULL,
	KEY_KP_1 = const.KEY_KP_1,
	KEY_KB_MENU = const.KEY_KB_MENU,
	MOUSE_CURSOR_DEFAULT = const.MOUSE_CURSOR_DEFAULT,
	MAROON = const.MAROON,
	KEY_APOSTROPHE = const.KEY_APOSTROPHE,
	KEY_F10 = const.KEY_F10,
	KEY_I = const.KEY_I,
	KEY_F9 = const.KEY_F9,
	KEY_RIGHT_BRACKET = const.KEY_RIGHT_BRACKET,
	KEY_F5 = const.KEY_F5,
	KEY_ENTER = const.KEY_ENTER,
	KEY_SLASH = const.KEY_SLASH,
	ptrs = const.ptrs,
	KEY_END = const.KEY_END,
	FLAG_INTERLACED_HINT = const.FLAG_INTERLACED_HINT,
	BLEND_ADDITIVE = const.BLEND_ADDITIVE,
	ORANGE = const.ORANGE,
	GESTURE_SWIPE_DOWN = const.GESTURE_SWIPE_DOWN,
	KEY_F7 = const.KEY_F7,
	Shader = const.Shader,
	GAMEPAD_BUTTON_RIGHT_TRIGGER_2 = const.GAMEPAD_BUTTON_RIGHT_TRIGGER_2,
	KEY_V = const.KEY_V,
	KEY_Q = const.KEY_Q,
	KEY_N = const.KEY_N,
	KEY_M = const.KEY_M,
	CUBEMAP_LAYOUT_AUTO_DETECT = const.CUBEMAP_LAYOUT_AUTO_DETECT,
	KEY_F = const.KEY_F,
	KEY_EQUAL = const.KEY_EQUAL,
	SHADER_ATTRIB_VEC4 = const.SHADER_ATTRIB_VEC4,
	KEY_W = const.KEY_W,
	BEIGE = const.BEIGE,
	KEY_ONE = const.KEY_ONE,
	KEY_NUM_LOCK = const.KEY_NUM_LOCK,
	KEY_COMMA = const.KEY_COMMA,
	PIXELFORMAT_UNCOMPRESSED_GRAYSCALE = const.PIXELFORMAT_UNCOMPRESSED_GRAYSCALE,
	LOG_ALL = const.LOG_ALL,
	ModelAnimation = const.ModelAnimation,
	KEY_DOWN = const.KEY_DOWN,
	KEY_KP_SUBTRACT = const.KEY_KP_SUBTRACT,
	GAMEPAD_AXIS_LEFT_TRIGGER = const.GAMEPAD_AXIS_LEFT_TRIGGER,
	BLUE = const.BLUE,
	KEY_LEFT_CONTROL = const.KEY_LEFT_CONTROL,
	KEY_ESCAPE = const.KEY_ESCAPE,
	KEY_KP_7 = const.KEY_KP_7,
}
