--[=[
-- Used this script to dump Roblox's BrickColors
-- also found out while doing this, Roblox actually made duplicate names of some BrickColors

local str = [[local Color3 = require("@Color3")

local function Map(name, number, r, g, b)
	return {
		Name = name,
		Number = number,
		r = r,
		g = g,
		b = b
	}
end

return {
]]

for i = 1, 1024 do
	local success, brickColor = pcall(function()
		return BrickColor.new(i)
	end)
	
	if success and brickColor and (brickColor.Number == i) then

		print(i)
		str ..= ("    Map(\"%s\", %i, %i, %i, %i),\n"):format(brickColor.Name, i, math.round(brickColor.r*255), math.round(brickColor.g*255), math.round(brickColor.b*255))
	else
		warn(i)
	end
end

str ..= "}"

local m = Instance.new("ModuleScript")
m.Source = str
m.Name = "BrickColor"
m.Parent = workspace
]=]


local Color3 = require("@Color3")

local function Map(name, number, r, g, b)
	return {
		Name = name,
		Number = number,
		r = r,
		g = g,
		b = b
	}
end

return {
    Map("White", 1, 242, 243, 243),
    Map("Grey", 2, 161, 165, 162),
    Map("Light yellow", 3, 249, 233, 153),
    Map("Brick yellow", 5, 215, 197, 154),
    Map("Light green (Mint)", 6, 194, 218, 184),
    Map("Light reddish violet", 9, 232, 186, 200),
    Map("Pastel Blue", 11, 128, 187, 219),
    Map("Light orange brown", 12, 203, 132, 66),
    Map("Nougat", 18, 204, 142, 105),
    Map("Bright red", 21, 196, 40, 28),
    Map("Med. reddish violet", 22, 196, 112, 160),
    Map("Bright blue", 23, 13, 105, 172),
    Map("Bright yellow", 24, 245, 205, 48),
    Map("Earth orange", 25, 98, 71, 50),
    Map("Black", 26, 27, 42, 53),
    Map("Dark grey", 27, 109, 110, 108),
    Map("Dark green", 28, 40, 127, 71),
    Map("Medium green", 29, 161, 196, 140),
    Map("Lig. Yellowich orange", 36, 243, 207, 155),
    Map("Bright green", 37, 75, 151, 75),
    Map("Dark orange", 38, 160, 95, 53),
    Map("Light bluish violet", 39, 193, 202, 222),
    Map("Transparent", 40, 236, 236, 236),
    Map("Tr. Red", 41, 205, 84, 75),
    Map("Tr. Lg blue", 42, 193, 223, 240),
    Map("Tr. Blue", 43, 123, 182, 232),
    Map("Tr. Yellow", 44, 247, 241, 141),
    Map("Light blue", 45, 180, 210, 228),
    Map("Tr. Flu. Reddish orange", 47, 217, 133, 108),
    Map("Tr. Green", 48, 132, 182, 141),
    Map("Tr. Flu. Green", 49, 248, 241, 132),
    Map("Phosph. White", 50, 236, 232, 222),
    Map("Light red", 100, 238, 196, 182),
    Map("Medium red", 101, 218, 134, 122),
    Map("Medium blue", 102, 110, 153, 202),
    Map("Light grey", 103, 199, 193, 183),
    Map("Bright violet", 104, 107, 50, 124),
    Map("Br. yellowish orange", 105, 226, 155, 64),
    Map("Bright orange", 106, 218, 133, 65),
    Map("Bright bluish green", 107, 0, 143, 156),
    Map("Earth yellow", 108, 104, 92, 67),
    Map("Bright bluish violet", 110, 67, 84, 147),
    Map("Tr. Brown", 111, 191, 183, 177),
    Map("Medium bluish violet", 112, 104, 116, 172),
    Map("Tr. Medi. reddish violet", 113, 229, 173, 200),
    Map("Med. yellowish green", 115, 199, 210, 60),
    Map("Med. bluish green", 116, 85, 165, 175),
    Map("Light bluish green", 118, 183, 215, 213),
    Map("Br. yellowish green", 119, 164, 189, 71),
    Map("Lig. yellowish green", 120, 217, 228, 167),
    Map("Med. yellowish orange", 121, 231, 172, 88),
    Map("Br. reddish orange", 123, 211, 111, 76),
    Map("Bright reddish violet", 124, 146, 57, 120),
    Map("Light orange", 125, 234, 184, 146),
    Map("Tr. Bright bluish violet", 126, 165, 165, 203),
    Map("Gold", 127, 220, 188, 129),
    Map("Dark nougat", 128, 174, 122, 89),
    Map("Silver", 131, 156, 163, 168),
    Map("Neon orange", 133, 213, 115, 61),
    Map("Neon green", 134, 216, 221, 86),
    Map("Sand blue", 135, 116, 134, 157),
    Map("Sand violet", 136, 135, 124, 144),
    Map("Medium orange", 137, 224, 152, 100),
    Map("Sand yellow", 138, 149, 138, 115),
    Map("Earth blue", 140, 32, 58, 86),
    Map("Earth green", 141, 39, 70, 45),
    Map("Tr. Flu. Blue", 143, 207, 226, 247),
    Map("Sand blue metallic", 145, 121, 136, 161),
    Map("Sand violet metallic", 146, 149, 142, 163),
    Map("Sand yellow metallic", 147, 147, 135, 103),
    Map("Dark grey metallic", 148, 87, 88, 87),
    Map("Black metallic", 149, 22, 29, 50),
    Map("Light grey metallic", 150, 171, 173, 172),
    Map("Sand green", 151, 120, 144, 130),
    Map("Sand red", 153, 149, 121, 119),
    Map("Dark red", 154, 123, 46, 47),
    Map("Tr. Flu. Yellow", 157, 255, 246, 123),
    Map("Tr. Flu. Red", 158, 225, 164, 194),
    Map("Gun metallic", 168, 117, 108, 98),
    Map("Red flip/flop", 176, 151, 105, 91),
    Map("Yellow flip/flop", 178, 180, 132, 85),
    Map("Silver flip/flop", 179, 137, 135, 136),
    Map("Curry", 180, 215, 169, 75),
    Map("Fire Yellow", 190, 249, 214, 46),
    Map("Flame yellowish orange", 191, 232, 171, 45),
    Map("Reddish brown", 192, 105, 64, 40),
    Map("Flame reddish orange", 193, 207, 96, 36),
    Map("Medium stone grey", 194, 163, 162, 165),
    Map("Royal blue", 195, 70, 103, 164),
    Map("Dark Royal blue", 196, 35, 71, 139),
    Map("Bright reddish lilac", 198, 142, 66, 133),
    Map("Dark stone grey", 199, 99, 95, 98),
    Map("Lemon metalic", 200, 130, 138, 93),
    Map("Light stone grey", 208, 229, 228, 223),
    Map("Dark Curry", 209, 176, 142, 68),
    Map("Faded green", 210, 112, 149, 120),
    Map("Turquoise", 211, 121, 181, 181),
    Map("Light Royal blue", 212, 159, 195, 233),
    Map("Medium Royal blue", 213, 108, 129, 183),
    Map("Rust", 216, 144, 76, 42),
    Map("Brown", 217, 124, 92, 70),
    Map("Reddish lilac", 218, 150, 112, 159),
    Map("Lilac", 219, 107, 98, 155),
    Map("Light lilac", 220, 167, 169, 206),
    Map("Bright purple", 221, 205, 98, 152),
    Map("Light purple", 222, 228, 173, 200),
    Map("Light pink", 223, 220, 144, 149),
    Map("Light brick yellow", 224, 240, 213, 160),
    Map("Warm yellowish orange", 225, 235, 184, 127),
    Map("Cool yellow", 226, 253, 234, 141),
    Map("Dove blue", 232, 125, 187, 221),
    Map("Medium lilac", 268, 52, 43, 117),
    Map("Slime green", 301, 80, 109, 84),
    Map("Smoky grey", 302, 91, 93, 105),
    Map("Dark blue", 303, 0, 16, 176),
    Map("Parsley green", 304, 44, 101, 29),
    Map("Steel blue", 305, 82, 124, 174),
    Map("Storm blue", 306, 51, 88, 130),
    Map("Lapis", 307, 16, 42, 220),
    Map("Dark indigo", 308, 61, 21, 133),
    Map("Sea green", 309, 52, 142, 64),
    Map("Shamrock", 310, 91, 154, 76),
    Map("Fossil", 311, 159, 161, 172),
    Map("Mulberry", 312, 89, 34, 89),
    Map("Forest green", 313, 31, 128, 29),
    Map("Cadet blue", 314, 159, 173, 192),
    Map("Electric blue", 315, 9, 137, 207),
    Map("Eggplant", 316, 123, 0, 123),
    Map("Moss", 317, 124, 156, 107),
    Map("Artichoke", 318, 138, 171, 133),
    Map("Sage green", 319, 185, 196, 177),
    Map("Ghost grey", 320, 202, 203, 209),
    Map("Lilac", 321, 167, 94, 155),
    Map("Plum", 322, 123, 47, 123),
    Map("Olivine", 323, 148, 190, 129),
    Map("Laurel green", 324, 168, 189, 153),
    Map("Quill grey", 325, 223, 223, 222),
    Map("Crimson", 327, 151, 0, 0),
    Map("Mint", 328, 177, 229, 166),
    Map("Baby blue", 329, 152, 194, 219),
    Map("Carnation pink", 330, 255, 152, 220),
    Map("Persimmon", 331, 255, 89, 89),
    Map("Maroon", 332, 117, 0, 0),
    Map("Gold", 333, 239, 184, 56),
    Map("Daisy orange", 334, 248, 217, 109),
    Map("Pearl", 335, 231, 231, 236),
    Map("Fog", 336, 199, 212, 228),
    Map("Salmon", 337, 255, 148, 148),
    Map("Terra Cotta", 338, 190, 104, 98),
    Map("Cocoa", 339, 86, 36, 36),
    Map("Wheat", 340, 241, 231, 199),
    Map("Buttermilk", 341, 254, 243, 187),
    Map("Mauve", 342, 224, 178, 208),
    Map("Sunrise", 343, 212, 144, 189),
    Map("Tawny", 344, 150, 85, 85),
    Map("Rust", 345, 143, 76, 42),
    Map("Cashmere", 346, 211, 190, 150),
    Map("Khaki", 347, 226, 220, 188),
    Map("Lily white", 348, 237, 234, 234),
    Map("Seashell", 349, 233, 218, 218),
    Map("Burgundy", 350, 136, 62, 62),
    Map("Cork", 351, 188, 155, 93),
    Map("Burlap", 352, 199, 172, 120),
    Map("Beige", 353, 202, 191, 163),
    Map("Oyster", 354, 187, 179, 178),
    Map("Pine Cone", 355, 108, 88, 75),
    Map("Fawn brown", 356, 160, 132, 79),
    Map("Hurricane grey", 357, 149, 137, 136),
    Map("Cloudy grey", 358, 171, 168, 158),
    Map("Linen", 359, 175, 148, 131),
    Map("Copper", 360, 150, 103, 102),
    Map("Medium brown", 361, 86, 66, 54),
    Map("Bronze", 362, 126, 104, 63),
    Map("Flint", 363, 105, 102, 92),
    Map("Dark taupe", 364, 90, 76, 66),
    Map("Burnt Sienna", 365, 106, 57, 9),
    Map("Institutional white", 1001, 248, 248, 248),
    Map("Mid gray", 1002, 205, 205, 205),
    Map("Really black", 1003, 17, 17, 17),
    Map("Really red", 1004, 255, 0, 0),
    Map("Deep orange", 1005, 255, 176, 0),
    Map("Alder", 1006, 180, 128, 255),
    Map("Dusty Rose", 1007, 163, 75, 75),
    Map("Olive", 1008, 193, 190, 66),
    Map("New Yeller", 1009, 255, 255, 0),
    Map("Really blue", 1010, 0, 0, 255),
    Map("Navy blue", 1011, 0, 32, 96),
    Map("Deep blue", 1012, 33, 84, 185),
    Map("Cyan", 1013, 4, 175, 236),
    Map("CGA brown", 1014, 170, 85, 0),
    Map("Magenta", 1015, 170, 0, 170),
    Map("Pink", 1016, 255, 102, 204),
    Map("Deep orange", 1017, 255, 175, 0),
    Map("Teal", 1018, 18, 238, 212),
    Map("Toothpaste", 1019, 0, 255, 255),
    Map("Lime green", 1020, 0, 255, 0),
    Map("Camo", 1021, 58, 125, 21),
    Map("Grime", 1022, 127, 142, 100),
    Map("Lavender", 1023, 140, 91, 159),
    Map("Pastel light blue", 1024, 175, 221, 255),
}